File: casu/export.py
"""Verified CASU extraction and transcoding."""
from __future__ import annotations

import os
import subprocess
import tempfile
import wave
from pathlib import Path

import numpy as np

from .core import CasuError, ffprobe, require_tool, resolve_casu_source
from .native import NativeCasuError, read_native
from .native_v2 import (ChunkType, NativeV2Error, TileStateCache,
                        decode_attachment, decode_audio_block, decode_chapter_table,
                        decode_bitmap_subtitle, decode_subtitle_packet,
                        read_native_v2)


class CasuExportError(CasuError):
    pass


_AUDIO_EXTENSIONS = {".aac", ".aiff", ".alac", ".flac", ".m4a", ".mka", ".mp3",
                     ".oga", ".ogg", ".opus", ".wav", ".wma"}


def _atomic_ffmpeg(command: list[str], destination: Path) -> Path:
    destination = destination.expanduser().resolve()
    destination.parent.mkdir(parents=True, exist_ok=True)
    fd, name = tempfile.mkstemp(prefix=f".{destination.stem}.",
                                suffix=destination.suffix or ".media",
                                dir=destination.parent)
    os.close(fd)
    temporary = Path(name)
    try:
        process = subprocess.run(command[:-1] + [str(temporary)], text=True,
                                 stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        if process.returncode:
            detail = process.stderr.strip().splitlines()
            raise CasuExportError(detail[-1] if detail else "FFmpeg export failed")
        if not temporary.is_file() or temporary.stat().st_size == 0:
            raise CasuExportError("FFmpeg produced an empty export")
        os.replace(temporary, destination)
        return destination
    finally:
        try:
            temporary.unlink()
        except FileNotFoundError:
            pass


def _codec_options(destination: Path, *, has_video: bool, has_audio: bool,
                   has_subtitles: bool = False,
                   has_rich_subtitles: bool = False) -> list[str]:
    extension = destination.suffix.lower()
    if not has_video:
        return {
            ".mp3": ["-c:a", "libmp3lame", "-q:a", "2"],
            ".flac": ["-c:a", "flac"],
            ".wav": ["-c:a", "pcm_s16le"],
            ".ogg": ["-c:a", "libvorbis", "-q:a", "6"],
            ".oga": ["-c:a", "libvorbis", "-q:a", "6"],
            ".opus": ["-c:a", "libopus", "-b:a", "160k"],
            ".m4a": ["-c:a", "aac", "-b:a", "192k"],
            ".aac": ["-c:a", "aac", "-b:a", "192k"],
        }.get(extension, [])
    video = {
        ".webm": ["-c:v", "libvpx-vp9", "-crf", "24", "-b:v", "0"],
        ".avi": ["-c:v", "mpeg4", "-q:v", "3"],
    }.get(extension, ["-c:v", "libx264", "-crf", "18", "-pix_fmt", "yuv420p"])
    if not has_audio:
        result = video
    else:
        audio = (["-c:a", "libopus", "-b:a", "160k"] if extension == ".webm"
                 else ["-c:a", "aac", "-b:a", "192k"])
        result = video + audio
    if has_subtitles:
        result += (["-c:s", "webvtt"] if extension == ".webm" else
                   ["-c:s", "mov_text"] if extension in {".mp4", ".m4v", ".mov"} else
                   ["-c:s", "ass"] if has_rich_subtitles else
                   ["-c:s", "srt"])
    return result


def _legacy_source(casu_path: Path, temporary_directory: Path) -> Path | None:
    with casu_path.open("rb") as handle:
        magic = handle.read(8)
    if magic == b"CASUNAT1":
        container = read_native(casu_path, verify_payload=True)
        source_name = Path(str(container.manifest.get("source", {}).get("filename")
                               or "payload.bin")).name
        if not source_name or source_name in {".", ".."}:
            raise CasuExportError("CASUNAT1 source filename is invalid")
        return container.extract_payload(temporary_directory / source_name)
    if magic == b"CASUNAT2":
        return None
    return resolve_casu_source(casu_path)


def _native_bitmap_subtitles(container, stream_ids: set[int]) -> list:
    packets = []
    for offset, summary in zip(container.offsets, container.chunks):
        if (summary.stream_id in stream_ids
                and summary.chunk_type == ChunkType.SUBTITLE_BITMAP):
            chunk, _following = container.read_chunk_at(offset)
            packets.append(decode_bitmap_subtitle(chunk.payload))
    return sorted(packets, key=lambda packet: (packet.start_pts, packet.end_pts))


def _burn_bitmap_subtitles(rgb: np.ndarray, seconds: float, packets: list) -> np.ndarray:
    """Alpha-composite active native bitmap cues into one reconstructed frame."""
    result = np.asarray(rgb, dtype=np.uint8).copy()
    frame_height, frame_width = result.shape[:2]
    milliseconds = seconds * 1000.0
    for packet in packets:
        if not packet.start_pts <= milliseconds < packet.end_pts:
            continue
        x0 = max(0, min(frame_width, round(packet.x * frame_width /
                                           packet.canvas_width)))
        y0 = max(0, min(frame_height, round(packet.y * frame_height /
                                            packet.canvas_height)))
        x1 = max(x0, min(frame_width, round((packet.x + packet.width) *
                                            frame_width / packet.canvas_width)))
        y1 = max(y0, min(frame_height, round((packet.y + packet.height) *
                                             frame_height / packet.canvas_height)))
        if x1 <= x0 or y1 <= y0:
            continue
        source = np.frombuffer(packet.rgba, dtype=np.uint8).reshape(
            packet.height, packet.width, 4)
        rows = np.minimum(packet.height - 1,
                          np.arange(y1 - y0) * packet.height // (y1 - y0))
        columns = np.minimum(packet.width - 1,
                             np.arange(x1 - x0) * packet.width // (x1 - x0))
        scaled = source[rows[:, None], columns[None, :]]
        alpha = scaled[:, :, 3:4].astype(np.uint16)
        target = result[y0:y1, x0:x1].astype(np.uint16)
        result[y0:y1, x0:x1] = ((scaled[:, :, :3].astype(np.uint16) * alpha
                                 + target * (255 - alpha) + 127) // 255).astype(np.uint8)
    return result


def _write_native_video(container, directory: Path, descriptor: dict,
                        bitmap_subtitles: list | None = None) -> Path:
    from mpcasu_native_backend import canonical_to_rgb

    stream_id = int(descriptor["stream_id"])
    timeline = descriptor.get("frame_timeline") or []
    if not timeline:
        raise CasuExportError("CASUNAT2 video stream has no frame timeline")
    relevant = []
    for offset, summary in zip(container.offsets, container.chunks):
        if (summary.stream_id == stream_id and summary.chunk_type in
                {ChunkType.VIDEO_KEY_STATE, ChunkType.VIDEO_TILE_UPDATE}):
            relevant.append((summary.pts, offset, summary.chunk_type))
    relevant.sort(key=lambda item: (item[0], item[1]))
    cache = TileStateCache()
    chunk_index = 0
    prefix = f"video-{stream_id}"
    concat = directory / f"{prefix}.ffconcat"
    lines = ["ffconcat version 1.0"]
    for index, entry in enumerate(timeline):
        pts = int(entry["pts"])
        while chunk_index < len(relevant) and relevant[chunk_index][0] <= pts:
            _chunk_pts, offset, kind = relevant[chunk_index]
            chunk, _following = container.read_chunk_at(offset)
            if kind == ChunkType.VIDEO_KEY_STATE:
                cache.apply_key_state(chunk.payload)
            else:
                cache.apply_tile_update(chunk.payload)
            chunk_index += 1
        if cache.frame is None:
            raise CasuExportError("CASUNAT2 timeline precedes its first key state")
        rgb = canonical_to_rgb(cache.frame)
        num, den = (int(value) for value in descriptor["time_base"])
        rgb = _burn_bitmap_subtitles(rgb, pts * num / den,
                                     bitmap_subtitles or [])
        height, width, _channels = rgb.shape
        frame_name = f"{prefix}-frame-{index:09d}.ppm"
        (directory / frame_name).write_bytes(
            f"P6\n{width} {height}\n255\n".encode("ascii") + rgb.tobytes())
        duration_pts = int(entry.get("duration_pts") or 0)
        duration = duration_pts * num / den
        lines.append(f"file '{frame_name}'")
        if duration > 0:
            lines.append(f"duration {duration:.12f}")
    lines.append(f"file '{prefix}-frame-{len(timeline) - 1:09d}.ppm'")
    concat.write_text("\n".join(lines) + "\n", encoding="utf-8")
    return concat


def _write_native_audio(container, directory: Path, descriptor: dict) -> Path:
    stream_id = int(descriptor["stream_id"])
    blocks = []
    for offset, summary in zip(container.offsets, container.chunks):
        if summary.stream_id == stream_id and summary.chunk_type == ChunkType.AUDIO_BLOCK:
            chunk, _following = container.read_chunk_at(offset)
            blocks.append(decode_audio_block(chunk.payload))
    if not blocks:
        raise CasuExportError("CASUNAT2 audio stream has no PCM blocks")
    blocks.sort(key=lambda block: block.pts * block.time_base_num / block.time_base_den)
    sample_rate, channels = blocks[0].sample_rate, blocks[0].channels
    target = directory / f"audio-{stream_id}.wav"
    cursor = 0
    with wave.open(str(target), "wb") as output:
        output.setnchannels(channels)
        output.setsampwidth(2)
        output.setframerate(sample_rate)
        for block in blocks:
            if (block.sample_rate, block.channels, block.sample_format) != (sample_rate, channels, "s16le"):
                raise CasuExportError("CASUNAT2 audio format changes are not exportable")
            start = round((block.pts * block.time_base_num / block.time_base_den) * sample_rate)
            if start > cursor:
                output.writeframesraw(b"\0" * ((start - cursor) * channels * 2))
                cursor = start
            skip = max(0, cursor - start)
            if skip < block.sample_count:
                output.writeframesraw(block.pcm[skip * channels * 2:])
                cursor += block.sample_count - skip
    return target


def _srt_time(milliseconds: int) -> str:
    value = max(0, int(milliseconds))
    hours, remainder = divmod(value, 3_600_000)
    minutes, remainder = divmod(remainder, 60_000)
    seconds, millis = divmod(remainder, 1000)
    return f"{hours:02d}:{minutes:02d}:{seconds:02d},{millis:03d}"


def _write_native_subtitle(container, directory: Path,
                           descriptor: dict) -> tuple[Path, bool] | None:
    stream_id = int(descriptor["stream_id"])
    if descriptor.get("rich_source_attachment"):
        rich = []
        for offset, summary in zip(container.offsets, container.chunks):
            if (summary.stream_id == stream_id and
                    summary.chunk_type == ChunkType.ATTACHMENT):
                chunk, _following = container.read_chunk_at(offset)
                attachment = decode_attachment(chunk.payload)
                if attachment.role == "subtitle-source":
                    rich.append(attachment)
        if len(rich) != 1:
            raise CasuExportError(
                "CASUNAT2 rich subtitle stream lacks one verified source attachment")
        target = directory / f"subtitle-{stream_id}.ass"
        target.write_bytes(rich[0].data)
        return target, True
    packets = []
    for offset, summary in zip(container.offsets, container.chunks):
        if (summary.stream_id == stream_id and
                summary.chunk_type == ChunkType.SUBTITLE_PACKET):
            chunk, _following = container.read_chunk_at(offset)
            packets.append(decode_subtitle_packet(chunk.payload))
    if not packets:
        return None
    packets.sort(key=lambda packet: (packet.start_pts, packet.end_pts))
    target = directory / f"subtitle-{stream_id}.srt"
    lines = []
    for index, packet in enumerate(packets, start=1):
        lines.extend([str(index),
                      f"{_srt_time(packet.start_pts)} --> {_srt_time(packet.end_pts)}",
                      packet.text, ""])
    target.write_text("\n".join(lines), encoding="utf-8")
    return target, False


def _ffmetadata_value(value: str) -> str:
    return (value.replace("\\", "\\\\").replace("\n", "\\\n")
            .replace("=", "\\=").replace(";", "\\;").replace("#", "\\#"))


def _write_native_chapters(container, directory: Path) -> Path | None:
    tables = []
    for offset, summary in zip(container.offsets, container.chunks):
        if summary.chunk_type == ChunkType.CHAPTER_TABLE:
            chunk, _following = container.read_chunk_at(offset)
            tables.append(decode_chapter_table(chunk.payload))
    if not tables:
        return None
    if len(tables) != 1:
        raise CasuExportError("CASUNAT2 contains multiple chapter tables")
    target = directory / "chapters.ffmetadata"
    lines = [";FFMETADATA1"]
    for chapter in tables[0]:
        lines.extend([
            "[CHAPTER]", "TIMEBASE=1/1000000000",
            f"START={int(chapter['start_pts'])}",
            f"END={int(chapter['end_pts'])}",
            f"title={_ffmetadata_value(str(chapter['title']))}",
            f"language={_ffmetadata_value(str(chapter.get('language') or 'und'))}",
        ])
    target.write_text("\n".join(lines) + "\n", encoding="utf-8")
    return target


def export_casu(source: str | Path, destination: str | Path) -> Path:
    """Verify and export any supported CASU representation to an FFmpeg format."""
    require_tool("ffmpeg")
    source_path = Path(source).expanduser().resolve()
    destination_path = Path(destination).expanduser().resolve()
    if not source_path.is_file():
        raise CasuExportError("export input must be an existing CASU file")
    if destination_path == source_path or not destination_path.suffix:
        raise CasuExportError("export destination must use a media-file extension")
    try:
        with tempfile.TemporaryDirectory(prefix="casu-export-") as name:
            directory = Path(name)
            legacy = _legacy_source(source_path, directory)
            if legacy is not None:
                is_audio = destination_path.suffix.lower() in _AUDIO_EXTENSIONS
                overview = ffprobe(legacy)
                stream_types = [item.get("codec_type")
                                for item in overview.get("streams", [])]
                has_video = "video" in stream_types and not is_audio
                has_audio = "audio" in stream_types
                has_subtitles = "subtitle" in stream_types and not is_audio
                command = ["ffmpeg", "-v", "error", "-y", "-i", str(legacy)]
                command += (["-map", "0:a:0"] if is_audio else [
                    "-map", "0:v?", "-map", "0:a?", "-map", "0:s?",
                    "-map_chapters", "0",
                ])
                command += _codec_options(destination_path, has_video=has_video,
                                          has_audio=has_audio,
                                          has_subtitles=has_subtitles)
                return _atomic_ffmpeg(command + [str(destination_path)], destination_path)

            container = read_native_v2(source_path, load_payloads=False)
            streams = [item for item in container.manifest.get("streams", [])
                       if isinstance(item, dict)]
            videos = [item for item in streams if item.get("type") == "video"]
            audios = [item for item in streams if item.get("type") == "audio"]
            subtitles = [item for item in streams if item.get("type") == "subtitle"]
            if destination_path.suffix.lower() in _AUDIO_EXTENSIONS:
                videos = []
                audios = audios[:1]
                subtitles = []
            if not videos and not audios:
                raise CasuExportError("CASUNAT2 has no exportable audio/video stream")
            inputs: list[str] = []
            maps: list[str] = []
            input_index = 0
            bitmap_descriptors = [
                item for item in subtitles
                if (item.get("canonical_format") == "rgba-bitmap-region"
                    or str(item.get("codec_origin") or "").lower() in {
                        "hdmv_pgs_subtitle", "dvd_subtitle", "dvb_subtitle", "xsub"
                    })
            ]
            # A flattened video can contain one visible subtitle language. Honour
            # the container default, otherwise use the first stable stream order.
            selected_bitmap = next(
                (item for item in bitmap_descriptors
                 if item.get("default") or item.get("disposition", {}).get("default")),
                bitmap_descriptors[0] if bitmap_descriptors else None,
            )
            bitmap_streams = ({int(selected_bitmap["stream_id"])}
                              if selected_bitmap is not None else set())
            bitmap_packets = _native_bitmap_subtitles(container, bitmap_streams)
            for video in videos:
                inputs += ["-f", "concat", "-safe", "0", "-i",
                           str(_write_native_video(container, directory, video,
                                                   bitmap_packets))]
                maps += ["-map", f"{input_index}:v:0"]
                input_index += 1
            for audio in audios:
                inputs += ["-i", str(_write_native_audio(container, directory, audio))]
                maps += ["-map", f"{input_index}:a:0"]
                input_index += 1
            written_subtitles = []
            has_rich_subtitles = False
            for subtitle in subtitles:
                written = _write_native_subtitle(container, directory, subtitle)
                if written is None:
                    continue
                subtitle_path, rich = written
                inputs += ["-i", str(subtitle_path)]
                maps += ["-map", f"{input_index}:s:0"]
                written_subtitles.append(subtitle_path)
                has_rich_subtitles = has_rich_subtitles or rich
                input_index += 1
            chapters = _write_native_chapters(container, directory)
            chapter_input = None
            if chapters is not None:
                inputs += ["-f", "ffmetadata", "-i", str(chapters)]
                chapter_input = input_index
                input_index += 1
            command = ["ffmpeg", "-v", "error", "-y"] + inputs + maps
            if chapter_input is not None:
                command += ["-map_chapters", str(chapter_input)]
            command += _codec_options(destination_path, has_video=bool(videos),
                                      has_audio=bool(audios),
                                      has_subtitles=bool(written_subtitles),
                                      has_rich_subtitles=has_rich_subtitles)
            return _atomic_ffmpeg(command + [str(destination_path)], destination_path)
    except (OSError, ValueError, CasuError, NativeCasuError, NativeV2Error) as exc:
        if isinstance(exc, CasuExportError):
            raise
        raise CasuExportError(f"CASU export failed: {exc}") from exc

File: win-release/src/codec/casu/codec/export.cpp
// SPDX-License-Identifier: LicenseRef-CASU-AntiCapitalist-1.4
#include "casu/codec/export.hpp"

#include "casu/codec/ffmpeg.hpp"
#include "casu/codec/ffprobe.hpp"
#include "casu/formats.hpp"
#include "casu/json.hpp"
#include "casu/mp5.hpp"
#include "casu/native.hpp"
#include "casu/sidecar.hpp"

#include <chrono>
#include <cstdio>
#include <filesystem>
#include <map>
#include <random>
#include <set>
#include <string>
#include <vector>

namespace casu::codec {

namespace {

const std::set<std::string>& audio_extensions() {
    static const std::set<std::string> value = {
        ".aac", ".aiff", ".alac", ".flac", ".m4a", ".mka", ".mp3",
        ".oga", ".ogg", ".opus", ".wav", ".wma",
    };
    return value;
}

std::string lowercase_extension(const std::string& path) {
    std::string ext = std::filesystem::path(path).extension().string();
    for (char& c : ext) {
        if (c >= 'A' && c <= 'Z') c = char(c + ('a' - 'A'));
    }
    return ext;
}

std::string unique_temp_path(const std::filesystem::path& dir, const std::string& prefix,
                             const std::string& suffix) {
    static std::mt19937_64 gen(std::random_device{}());
    for (int attempt = 0; attempt < 16; ++attempt) {
        const std::filesystem::path candidate =
            dir / (prefix + std::to_string(gen()) + suffix);
        std::error_code ec;
        if (!std::filesystem::exists(candidate, ec)) return candidate.string();
    }
    throw CasuExportError("could not allocate a temporary path in " + dir.string());
}

void remove_file(const std::string& path) {
    std::error_code ec;
    std::filesystem::remove(path, ec);
}

std::vector<std::string> codec_options(const std::string& destination, bool has_video,
                                       bool has_audio, bool has_subtitles,
                                       bool has_rich_subtitles) {
    if (!has_video) {
        static const std::map<std::string, std::vector<std::string>> audio_table = {
            {".mp3", {"-c:a", "libmp3lame", "-q:a", "2"}},
            {".flac", {"-c:a", "flac"}},
            {".wav", {"-c:a", "pcm_s16le"}},
            {".ogg", {"-c:a", "libvorbis", "-q:a", "6"}},
            {".oga", {"-c:a", "libvorbis", "-q:a", "6"}},
            {".opus", {"-c:a", "libopus", "-b:a", "160k"}},
            {".m4a", {"-c:a", "aac", "-b:a", "192k"}},
            {".aac", {"-c:a", "aac", "-b:a", "192k"}},
        };
        const auto it = audio_table.find(lowercase_extension(destination));
        return it == audio_table.end() ? std::vector<std::string>{} : it->second;
    }
    const std::string ext = lowercase_extension(destination);
    std::vector<std::string> video =
        ext == ".webm" ? std::vector<std::string>{"-c:v", "libvpx-vp9", "-crf", "24", "-b:v", "0"}
        : ext == ".avi" ? std::vector<std::string>{"-c:v", "mpeg4", "-q:v", "3"}
                        : std::vector<std::string>{"-c:v", "libx264", "-crf", "18", "-pix_fmt", "yuv420p"};
    std::vector<std::string> result = video;
    if (has_audio) {
        const std::vector<std::string> audio =
            ext == ".webm" ? std::vector<std::string>{"-c:a", "libopus", "-b:a", "160k"}
                           : std::vector<std::string>{"-c:a", "aac", "-b:a", "192k"};
        result.insert(result.end(), audio.begin(), audio.end());
    }
    if (has_subtitles) {
        const std::vector<std::string> subtitle =
            ext == ".webm" ? std::vector<std::string>{"-c:s", "webvtt"}
            : (ext == ".mp4" || ext == ".m4v" || ext == ".mov")
                ? std::vector<std::string>{"-c:s", "mov_text"}
                : has_rich_subtitles ? std::vector<std::string>{"-c:s", "ass"}
                                     : std::vector<std::string>{"-c:s", "srt"};
        result.insert(result.end(), subtitle.begin(), subtitle.end());
    }
    return result;
}

bool path_has_stream(const JsonValue& probe, const std::string& codec_type) {
    const JsonValue* streams = probe.find("streams");
    if (!streams || !streams->is_array()) return false;
    for (const JsonValue& item : streams->as_array().items) {
        const JsonValue* type = item.find("codec_type");
        if (type && type->is_string() && type->as_string() == codec_type) return true;
    }
    return false;
}

std::string safe_source_name(const JsonValue& manifest) {
    const JsonValue* source = manifest.find("source");
    if (!source || !source->is_object()) throw CasuExportError("CASUNAT1 source is invalid");
    const JsonValue* filename = source->find("filename");
    if (!filename || !filename->is_string()) throw CasuExportError("CASUNAT1 source filename is invalid");
    const std::string raw = std::filesystem::path(filename->as_string()).filename().string();
    if (raw.empty() || raw == "." || raw == "..")
        throw CasuExportError("CASUNAT1 source filename is invalid");
    return raw;
}

// Extract the embedded original media from a supported CASU representation
// into a temporary file. Returns the temporary file path.
std::string legacy_source(const std::string& casu_path,
                          const std::filesystem::path& work_dir) {
    const casu::CasuKind kind = casu::detect_casu_kind(casu_path);
    if (kind == casu::CasuKind::Casunat1) {
        const casunat1::Container container = casunat1::read_native(casu_path, true);
        const std::string source_name = safe_source_name(container.manifest);
        const std::string out = (work_dir / source_name).string();
        container.extract_payload(out);
        return out;
    }
    if (kind == casu::CasuKind::Sidecar) return casu::resolve_casu_source(casu_path);
    if (kind == casu::CasuKind::Mp5) {
        const std::pair<std::string, std::vector<uint8_t>> attachment =
            casu::mp5::extract_attachment(casu_path);
        std::string filename = std::filesystem::path(attachment.first).filename().string();
        if (filename.empty()) filename = "payload.bin";
        const std::string out = (work_dir / filename).string();
        FILE* f = std::fopen(out.c_str(), "wb");
        if (!f) throw CasuExportError("could not create temp media: " + out);
        const std::vector<uint8_t>& data = attachment.second;
        const bool wrote = std::fwrite(data.data(), 1, data.size(), f) == data.size();
        std::fclose(f);
        if (!wrote) {
            remove_file(out);
            throw CasuExportError("could not write temp media: " + out);
        }
        return out;
    }
    if (kind == casu::CasuKind::Casunat2)
        throw CasuExportError(
            "CASUNAT2 native export requires the native decoder (planned; "
            "not yet implemented in this phase)");
    throw CasuExportError("input is not a recognised CASU representation");
}

// Run ffmpeg to a temporary file in the destination directory, verify the
// output, then atomically rename it over the destination.
std::string atomic_ffmpeg_impl(const std::vector<std::string>& args,
                               const std::string& destination,
                               bool append_output_arg) {
    const std::filesystem::path dest(destination);
    std::filesystem::path parent = dest.parent_path();
    if (parent.empty()) parent = ".";
    const std::string suffix = lowercase_extension(destination).empty() ? ".media" : dest.extension().string();
    const std::string temporary =
        unique_temp_path(parent, "." + dest.stem().string() + ".", suffix);
    std::vector<std::string> command = args;
    if (append_output_arg) command.push_back(temporary);
    else if (!command.empty()) command.back() = temporary;  // replace output arg
    try {
        Ffmpeg ffmpeg;
        ffmpeg.run_checked(command);
        std::error_code ec;
        if (!std::filesystem::exists(temporary, ec) ||
            std::filesystem::file_size(temporary, ec) == 0)
            throw CasuExportError("FFmpeg produced an empty export");
        JsonValue output_probe;
        try {
            output_probe = probe_json(temporary);
        } catch (const MediaProbeError& exc) {
            throw CasuExportError(std::string("exported output is unreadable: ") + exc.what());
        }
        if (!path_has_stream(output_probe, "video") && !path_has_stream(output_probe, "audio"))
            throw CasuExportError("converted output has no playable stream");
        remove_file(destination);
        std::filesystem::rename(temporary, dest, ec);
        if (ec) {
            remove_file(temporary);
            throw CasuExportError("could not finalize export: " + destination);
        }
        return destination;
    } catch (...) {
        remove_file(temporary);
        throw;
    }
}
}

// Public transcode_media parity entry point: args carry the FINAL output as
// last argument (build_transcode_command form); it is swapped for a temp path
// and published atomically after verification.
std::string transcode_atomic(const std::vector<std::string>& args,
                             const std::string& destination) {
    return atomic_ffmpeg_impl(args, destination, /*append_output_arg=*/false);
}

void export_casu(const std::string& source, const std::string& destination) {
    std::error_code ec;
    const std::filesystem::path source_abs = std::filesystem::absolute(source, ec).lexically_normal();
    const std::filesystem::path dest_abs = std::filesystem::absolute(destination, ec).lexically_normal();
    if (ec || !std::filesystem::is_regular_file(source_abs, ec))
        throw CasuExportError("export input must be an existing CASU file");
    if (dest_abs == source_abs || dest_abs.extension().empty())
        throw CasuExportError("export destination must use a media-file extension");

    std::filesystem::path work_dir;
    for (int attempt = 0; attempt < 8; ++attempt) {
        static std::mt19937_64 gen(std::random_device{}());
        work_dir = std::filesystem::temp_directory_path() /
                   ("casu-export-" + std::to_string(gen()));
        if (std::filesystem::create_directory(work_dir, ec)) break;
    }
    if (ec || !std::filesystem::exists(work_dir, ec))
        throw CasuExportError("could not create export working directory");

    try {
        const std::string legacy = legacy_source(source_abs.string(), work_dir);
        const bool is_audio = audio_extensions().count(lowercase_extension(dest_abs.string())) != 0;
        JsonValue overview;
        try {
            overview = probe_json(legacy);
        } catch (const MediaProbeError& exc) {
            throw CasuExportError(std::string("media probe failed: ") + exc.what());
        }
        const bool has_video = path_has_stream(overview, "video") && !is_audio;
        const bool has_audio = path_has_stream(overview, "audio");
        const bool has_subtitles = path_has_stream(overview, "subtitle") && !is_audio;

        std::vector<std::string> args = {"-v", "error", "-y", "-i", legacy};
        if (is_audio) {
            args.push_back("-map"), args.push_back("0:a:0");
        } else {
            args.push_back("-map"), args.push_back("0:v?");
            args.push_back("-map"), args.push_back("0:a?");
            args.push_back("-map"), args.push_back("0:s?");
            args.push_back("-map_chapters"), args.push_back("0");
        }
        const std::vector<std::string> options =
            codec_options(dest_abs.string(), has_video, has_audio, has_subtitles, false);
        args.insert(args.end(), options.begin(), options.end());
        atomic_ffmpeg_impl(args, dest_abs.string(), /*append_output_arg=*/true);
    } catch (...) {
        std::error_code cleanup;
        std::filesystem::remove_all(work_dir, cleanup);
        throw;
    }
    std::error_code cleanup;
    std::filesystem::remove_all(work_dir, cleanup);
}

}  // namespace casu::codec
